### PR TITLE
RPI 1113039 - Jenkins: Plugin configuration that enables/disables the plugin from running jobs

### PR DESCRIPTION
**Pull request by PoojaVishal**
ability within the AccuRev Jenkins Plug-in to be able to turn on/off (start/stop, disable/enable) the plugin from running any jobs.
I have added check box "Disable Plugin for this server" in config.groovy to turn off that particualar accurev server form building any jobs when checked. Basically i skip the checkout in AccurevSCM.java.
image

Whenever disabled server is selected for any job>configure>source code management>Accurev server dropdown, depot textbox sould be read-only indicating that this server is disabled.
image

so i put if condition in config.jelly and i am unsure of syntax i am using to call method that gives if server is on/off.

# i get compilation error:
[INFO] --- findbugs-maven-plugin:3.0.4:check (findbugs) @ accurev ---
[INFO] BugInstance size is 1
[INFO] Error size is 0
[INFO] Total bugs: 1
_[INFO] _Possible null pointer dereference in hudson.plugins.accurev.AccurevSCM$AccurevSCMDescriptor.showDepot(String) due to return value of called method [hudson.plugins.accurev.AccurevSCM$AccurevSCMDescriptor, hudson.plugins.accurev.AccurevSCM$AccurevSCMDescriptor] Dereferenced at AccurevSCM.java:[line 566]Known null at AccurevSCM.java:[line 565] NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE_
[INFO]_ 

To see bug detail using the Findbugs GUI, use the following command "mvn findbugs:gui"

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:42 min
[INFO] Finished at: 2018-03-22T15:08:00+05:30
[INFO] Final Memory: 75M/1132M

### CODE REVIEW

**Comment by casz** [comment]
Removing the server serves the same purpose. Or better yet just disable the Jenkins job under configure.

image
This is how you do it for an entire folder:
https://support.cloudbees.com/hc/en-us/articles/226478907-How-to-disable-all-jobs-in-a-specific-folder-

This defeats the whole purpose of the plugin. So why and what is this purpose of this function. Why would you want to implement this?

As it is right now this provides no value, at least in my mind. You can just as easily archive your jobs by removing them from the jobs folder inside the Jenkins server.

File: src/main/java/hudson/plugins/accurev/AccurevSCM.java [hunk 1]
```diff
@@ -877,8 +890,21 @@ public boolean isUsePromoteListen() {
...
+			return isServerDisabled;
+		}
+        
+        @DataBoundSetter
+		public void setServerDisabled(boolean isServerDisabled) {
+			this.isServerDisabled = isServerDisabled;
+		}
 
-        public boolean migrateCredentials() {
+		public boolean migrateCredentials() {
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/AccurevSCM.java]: Don't use tab, please install https://marketplace.eclipse.org/content/editorconfig-eclipse as it says spaces

File: src/main/java/hudson/plugins/accurev/AccurevSCM.java [hunk 2]
```diff
@@ -877,8 +890,21 @@ public boolean isUsePromoteListen() {
         public void setUsePromoteListen(boolean usePromoteListen) {
             this.usePromoteListen = usePromoteListen;
         }
+        
+      
+
+        
+       
+        public boolean isServerDisabled() {
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/AccurevSCM.java]: Please align properly, the method starts with spaces and ends with tabs...

File: src/main/java/hudson/plugins/accurev/AccurevSCM.java [hunk 3]
```diff
@@ -877,8 +890,21 @@ public boolean isUsePromoteListen() {
         public void setUsePromoteListen(boolean usePromoteListen) {
             this.usePromoteListen = usePromoteListen;
         }
+        
+      
+
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/AccurevSCM.java]: Remove any excess newlines only one single newline is allowed

File: src/main/java/hudson/plugins/accurev/AccurevSCM.java [hunk 4]
```diff
@@ -515,8 +521,9 @@ public String getPollingStream(Job<?, ?> project, TaskListener listener) throws
         // The transient modifier means it won't be written to the config file
         private transient List<AccurevServer> servers;
         private boolean pollOnMaster;
+        
 
-        /**
+		/**
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/AccurevSCM.java]: 💥 Tabs 💥

File: src/main/java/hudson/plugins/accurev/AccurevSCM.java [hunk 5]
```diff
@@ -366,14 +364,22 @@ public void buildEnvironment(Run<?, ?> build, Map<String, String> env) {
...
+		 boolean serverDisabled = server != null && server.isServerDisabled();
+		if (!serverDisabled) {
+			boolean checkout = AccurevMode.findDelegate(this).checkout(build, launcher, workspace, listener,
+					changelogFile);
+			if (checkout)
+				listener.getLogger().println("Checkout done");
+			else
+				listener.getLogger().println("Checkout failed");
+		} else {
+			listener.getLogger().println("Checkout skipped");
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/AccurevSCM.java]: This won't work... now your just breaking builds without proper checkout. Please make it throw `AbortException`.

File: src/main/java/hudson/plugins/accurev/AccurevSCM.java [hunk 6]
```diff
@@ -64,10 +63,9 @@
 import hudson.util.ListBoxModel;
 import hudson.util.Secret;
 import jenkins.model.Jenkins;
-
 import jenkins.plugins.accurev.AccurevTool;
 import jenkins.plugins.accurev.util.UUIDUtils;
-import hudson.plugins.accurev.delegates.AbstractModeDelegate;
+import net.sf.json.JSONObject;
```
Comment (casz) [inline src/main/java/hudson/plugins/accurev/AccurevSCM.java]: Please follow the current import order.

File: src/main/java/hudson/plugins/accurev/AccurevSCM.java (hunk 5 above)
Comment (casz) [reply src/main/java/hudson/plugins/accurev/AccurevSCM.java]: Please change it to the something like this instead.
```java
AccurevServer server = getServer();
boolean serverDisabled = server != null && server.isServerDisabled();
if (serverDisabled) {
  throw new AbortException("Checkout skipped");
}
```

**Comment by PoojaVishal** [comment]
Hi Joseph,

Thank you for your comments.

If i put throw new AbortException("Checkout skipped");
then there will be build failure. This will lead to trigger failed build event mails to organization which customer specially want to avoid. 

Regards,
Pooja


>>> Joseph Petersen <notifications@github.com> 3/27/2018 5:42 PM >>>
@casz commented on this pull request.
In src/main/java/hudson/plugins/accurev/AccurevSCM.java
( https://github.com/jenkinsci/accurev-plugin/pull/59#discussion_r177400892) :
> -        else listener.getLogger().println("Checkout failed");-    }+ public void checkout(@Nonnull Run<?, ?> build, @Nonnull Launcher launcher, @Nonnull FilePath workspace,+ @Nonnull TaskListener listener, @CheckForNull File changelogFile, @CheckForNull SCMRevisionState baseline)+ throws IOException, InterruptedException {+  AccurevServer server = getServer();+  boolean serverDisabled = server != null && server.isServerDisabled();+ if (!serverDisabled) {+ boolean checkout = AccurevMode.findDelegate(this).checkout(build, launcher, workspace, listener,+ changelogFile);+ if (checkout)+ listener.getLogger().println("Checkout done");+ else+ listener.getLogger().println("Checkout failed");+ } else {+ listener.getLogger().println("Checkout skipped");
Please change it to the something like this instead.
AccurevServer server = getServer();boolean serverDisabled = server != null && server.isServerDisabled();if (serverDisabled) { throw new AbortException("Checkout skipped");}
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
( https://github.com/jenkinsci/accurev-plugin/pull/59#discussion_r177400892) , or mute the thread

**Comment by casz** [comment]
Well without checkout the build will definitely still fail.

**Pull request closed without merging**